### PR TITLE
Validation rules skeleton (DEV)

### PR DESCRIPTION
**Pull request by mtwalli**
- Some skeleton classes for Validation rules feature

#### Note 
- Some data structure should be provided by the library 📚 used for JOSN engine
- covid_certificate_strings.xml should be used for localisation 
- covid_certificate_graph.xml should be used to add the new screens

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validationrules/rules/DgcRulesValidator.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+import de.rki.coronawarnapp.covidcertificate.common.repository.CertificateContainerId
+import de.rki.coronawarnapp.covidcertificate.test.core.TestCertificateRepository
+import de.rki.coronawarnapp.covidcertificate.validationrules.country.DgcCountry
+import org.joda.time.DateTime
+import javax.inject.Inject
+
+class DgcRulesValidator @Inject constructor(
+    testCertificateRepository: TestCertificateRepository,
+    dgcDgcValidationRulesRepository: DgcValidationRulesRepository,
```
Comment (axelherbstreith) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validationrules/rules/DgcRulesValidator.kt:11]: small typo

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validationrules/rules/DgcValidationResultType.kt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package de.rki.coronawarnapp.covidcertificate.validationrules.rules
+
+enum class DgcValidationResultType {
+    PASS,
+    FAIL,
+    OPEN
+}
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validationrules/rules/DgcValidationResultType.kt:7]: PASSED and FAILED as it describes a state

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validationrules/rules/CWT.kt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package de.rki.coronawarnapp.covidcertificate.validationrules.rules
+
+enum class CWT(val value: String) {
+    ISS("iss"),
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validationrules/rules/CWT.kt:4]: Not sure what this is for

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/ui/validationrules/validatoinresult/pass/DgcValidationPassFragment.kt [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package de.rki.coronawarnapp.covidcertificate.ui.validationrules.validatoinresult.pass
+
+import androidx.fragment.app.Fragment
+
+class DgcValidationPassFragment : Fragment()
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/ui/validationrules/validatoinresult/pass/DgcValidationPassFragment.kt:5]: Passed

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/ui/validationrules/validatoinresult/fail/DgcValidationFailFragment.kt [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package de.rki.coronawarnapp.covidcertificate.ui.validationrules.validatoinresult.fail
+
+import androidx.fragment.app.Fragment
+
+class DgcValidationFailFragment : Fragment()
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/ui/validationrules/validatoinresult/fail/DgcValidationFailFragment.kt:5]: Failed

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validationrules/rules/DgcRulesValidator.kt [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+) {
+
+    /**
+     * Validates DGC against country of arrival's rules and issuer country rules
+     */
+    suspend fun validateDgc(
+        arrivalsCountry: Set<DgcCountry>, // For future allow multiple country selection
+        containerId: CertificateContainerId,
+        dateTime: DateTime,
+        cwt: CWT
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validationrules/rules/DgcRulesValidator.kt:21]: This needs to be the DccHeader

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validationrules/rules/DgcRulesValidator.kt [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+class DgcRulesValidator @Inject constructor(
+    testCertificateRepository: TestCertificateRepository,
+    dgcDgcValidationRulesRepository: DgcValidationRulesRepository,
+) {
+
+    /**
+     * Validates DGC against country of arrival's rules and issuer country rules
+     */
+    suspend fun validateDgc(
+        arrivalsCountry: Set<DgcCountry>, // For future allow multiple country selection
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validationrules/rules/DgcRulesValidator.kt:18]: arrivalCountries

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validationrules/rules/DgcRulesValidator.kt [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        containerId: CertificateContainerId,
+        dateTime: DateTime,
+        cwt: CWT
+    ): DgcValidationResult {
+        return object : DgcValidationResult {
+            override val expirationCheckPassed: Boolean
+                get() = false
+            override val jsonSchemaCheckPassed: Boolean
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validationrules/rules/DgcRulesValidator.kt:26]: Not sure about the sensibility here - qr codes with invalid json won't even get here. They are rejected when scanning/ retrieved if invalid.

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/covidcertificate/vaccination/core/VaccinationTestData.kt [hunk 9]
```diff
@@ -173,7 +175,8 @@ class VaccinationTestData @Inject constructor(
 
     val personBVac1CertificateData = DccData(
         certificate = personBVac1Certificate,
-        header = personBVac1CertificateHeader
+        header = personBVac1CertificateHeader,
+        certificateJson = "{\"v\":[{\"ci\":\"01DE/00001/1119305005/3H24U2KVOTPCSINK7N64F2OB9#S\",\"co\":\"DE\",\"dn\":1,\"dt\":\"2021-04-20\",\"is\":\"Bundesministerium für Gesundheit - Test01\",\"ma\":\"ORG-100001417\",\"mp\":\"EU/1/20/1525\",\"sd\":1,\"tg\":\"840539006\",\"vp\":\"1119305005\"}],\"dob\":\"1966-11-11\",\"nam\":{\"fn\":\"Johnson Gültig\",\"gn\":\"Boris\",\"fnt\":\"JOHNSON<GUELTIG\",\"gnt\":\"BORIS\"},\"ver\":\"1.0.0\"}",
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/test/java/de/rki/coronawarnapp/covidcertificate/vaccination/core/VaccinationTestData.kt:179]: With """  you can avoid escaping everything

Comment (d4rken) [reply Corona-Warn-App/src/test/java/de/rki/coronawarnapp/covidcertificate/vaccination/core/VaccinationTestData.kt:179]: Good point, forgot about that yesterday.
I just copy pasted it into the `""` and Android Studio did all the escaping for me :grinning:, I'd leave it like this for now, if that is okay.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validationrules/rules/DgcRulesValidator.kt (hunk 8 above)
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validationrules/rules/DgcRulesValidator.kt]: I agree, see discussions on internal teams channel.
Let's keep it like this for now and refactor after the discussion has concluded.

**Review by chiljamgossow — APPROVED** [review]
Ok, let's merge and go from there

**Review by chris-cwa — APPROVED** [review]

**Pull request merged**